pkg/controller/certificates/issuing/internal/secret.go

```diff
@@ -249,15 +249,25 @@ func (s *SecretsManager) getCertificateSecret(ctx context.Context, crt *cmapi.Ce
 func (s *SecretsManager) setKeystores(crt *cmapi.Certificate, secret *corev1.Secret, data SecretData) error {
 	// Handle the experimental PKCS12 support
 	if crt.Spec.Keystores != nil && crt.Spec.Keystores.PKCS12 != nil && crt.Spec.Keystores.PKCS12.Create {
+		pw := []byte{}
 		ref := crt.Spec.Keystores.PKCS12.PasswordSecretRef
-		pwSecret, err := s.secretLister.Secrets(crt.Namespace).Get(ref.Name)
-		if err != nil {
-			return fmt.Errorf("fetching PKCS12 keystore password from Secret: %v", err)
-		}
-		if pwSecret.Data == nil || len(pwSecret.Data[ref.Key]) == 0 {
-			return fmt.Errorf("PKCS12 keystore password Secret contains no data for key %q", ref.Key)
+		if ref != nil {
+			pwSecret, err := s.secretLister.Secrets(crt.Namespace).Get(ref.Name)
+			if err != nil {
+				return fmt.Errorf("fetching PKCS12 keystore password from Secret: %v", err)
+			}
+			if pwSecret.Data == nil || len(pwSecret.Data[ref.Key]) == 0 {
+				return fmt.Errorf("PKCS12 keystore password Secret contains no data for key %q", ref.Key)
+			}
+			pw = pwSecret.Data[ref.Key]
+		} else if crt.Spec.Keystores.PKCS12.Password != nil {
+			pw = []byte(*crt.Spec.Keystores.PKCS12.Password)
+		} else {
+			return fmt.Errorf("either PasswordSecretRef or Password is mandatory")
 		}
-		pw := pwSecret.Data[ref.Key]
+		// Store the keystore password in the secret to help transition from PasswordSecretRef to Password
+		secret.Data[cmapi.KeystorePassword] = pw
```

Comment:
Some users of cert-manager might not like this. Maybe we should put this new feature behind a feature flag for now?

Reply:
To keep things simple, we could maybe put the password in the secret **only** in case of the new literal password ?
The idea behind this is to hardcode the password only in one place (the Certificate), and whoever is using the generated keystore keeps reading it dynamically, to avoid any risk of discrepency when hardcoding in 2 places.